Add support for Child Themes

jaxwilko opened the pull request:
This PR adds support for a theme to extend another theme via a `parent: [themeName]` directive in the `theme.yaml`. This PR also supports the `theme.yaml` being added as a `DbDatasource` content meaning that an extending theme can be purely driven by the DB.

Storm PR: https://github.com/wintercms/storm/pull/123

jaxwilko committed:
Added initial support for allowing a theme to extend a parent theme

jaxwilko committed:
Added an assetUrl helper for theme

jaxwilko committed:
Switched asset url generation to rely on theme assetUrl method

jaxwilko committed:
Added early exit if theme directory not exists

jaxwilko committed:
Added fix for loading items at the top of the filesystem

jaxwilko committed:
Switched order of execution

mjauvin commented:
@jaxwilko Do we want this PR for 1.2.1 since it's still a draft?

jaxwilko commented:
@mjauvin still got some bugs that need to be fixed before it's ready for review :)

mjauvin commented:
So it's mostly done, just has a few glitches?

LukeTowers commented:
@mjauvin this functionality is needed for our project that we're launching in November so what release it's tagged for doesn't really matter since it'll be merged when it's ready. Don't consider it blocking against 1.2.1 though.

LukeTowers committed:
Support child themes referencing their parent theme's localization strings

LukeTowers committed:
Add initial support for child themes referencing combined assets in parent themes

LukeTowers committed:
Avoid cache conflicts when running multiple themes

Instead of using 4 characters split across two folders for avoiding cache conflicts just use the relative path to the file directly. Also makes it easier to find files in the cache when debugging

LukeTowers committed:
Remove unintentional debugging statement

LukeTowers committed:
Avoid Twig cache conflicts when running multiple themes

Related: https://github.com/wintercms/winter/commit/a18e3ed112c3bfb98272d6dfc3c7c3b9de3192b2

jaxwilko committed:
Added file detection allowing per file based theme overrides

LukeTowers committed:
Set pivot data when initially syncing the relationship

Also performs the sync with model instances rather than just raw IDs to support relationships with a custom relatedKey set (the key on the related model's table that is stored on the pivot table to connect the relationship, normally just the primary key but can be anything).

LukeTowers committed:
Merge commit '966edea734a0e330f19ebe2547a1332ed39cb907' into wip/child-themes-luke

* commit '966edea734a0e330f19ebe2547a1332ed39cb907':
  Set pivot data when initially syncing the relationship
  Update jobs tables to support Laravel 9 (#730)
  Add migrate to list of protected commands (#733)
  Add "Send password reset email" button to backend users update page (#723)
  Adjust descriptor for local event
  Change event doc to test new event functionality in Docs plugin
  Removed NPX from webpack bin call (#728)
  Add test case for getParentData method in AJAX framework
  Recompile Snowboard
  Allow for custom AJAX error responses to be passed through handlers
  Allow detached AJAX request to be called with 2 params
  Improve IDE knowledge of the PluginBase object
  Fix support for data-request-parent

LukeTowers committed:
Merge branch 'develop' into wip/child-themes-luke

* develop:
  Set pivot data when initially syncing the relationship (#739)
  Added CLI signature to mix:list command
  Fix Navigation Manager unit test

LukeTowers committed:
Improve databaseLayerEnabled() check

LukeTowers committed:
Coding style fix

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Set pivot data when initially syncing the relationship (#739)
  Added CLI signature to mix:list command
  Fix Navigation Manager unit test
  Update jobs tables to support Laravel 9 (#730)
  Add migrate to list of protected commands (#733)
  Add "Send password reset email" button to backend users update page (#723)
  Adjust descriptor for local event
  Change event doc to test new event functionality in Docs plugin
  Removed NPX from webpack bin call (#728)
  Add test case for getParentData method in AJAX framework
  Recompile Snowboard
  Allow for custom AJAX error responses to be passed through handlers
  Allow detached AJAX request to be called with 2 params
  Improve IDE knowledge of the PluginBase object
  Fix support for data-request-parent

LukeTowers committed:
Merge branch 'wip/child-theme' into wip/child-themes-luke

LukeTowers committed:
Merge pull request #732 from wintercms/wip/child-themes-luke

Child Theme Improvements

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Switch from using Request::getBasePath() to Url::asset() (#742)

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Add missing import

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Fix base path for FrameworkNode

LukeTowers committed:
Include parent theme in twig caching path

jaxwilko committed:
Merge branch 'develop' into wip/child-theme

jaxwilko committed:
Moved backend localzation into theme activation to allow support for loading on theme activation instead of during service provider registration

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Allow Snowboard debugging to be controlled independently
  Show required indicators on checkbox fields (#754)
  Allow component assets to be mirrored by "winter:mirror" command

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Pass element's maxlength to slugify and toCamel to respect field definition (#762)

LukeTowers committed:
Improve performance of active theme resolution, add type hinting

This builds on the work originally done in https://github.com/octobercms/october/pull/3220 to further improve performance, especially in cases where lots of themes are present to choose from or "virtual" themes are being utilized.

Previously every single request would iterate over and initialize every single theme which involved booting an autodatasource for each theme and loading the entire contents of the path cache for each datasource of each theme into memory.

This commit changes the logic so that the cms.theme.getActiveTheme event will be run first, falling back to asking the cache service, then falling back to the DB if present, and finally defaulting to the value set in the configuration. After it has resolved an active theme it will attempt to cache that resolution permanently which should be fine since changing the active theme in the DB would trigger a cache invalidation when done through the setActiveTheme() method and changing the active theme in the file configuration would trigger a cache invalidation as well.

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Prevent fancy layout styling from spilling over into nested forms (#763)
  Allow Snowboard event listeners to be closures
  CMS Twig extension `contentFunction()` should return bool like `partialFunction()` (#746)
  prevent previous relations from being unlinked (#765)
  Allow cms.restrictBaseDir to be controlled via env variable

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Use the Storm File model's getLocalRootPath() implementation

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Ensure backend permissions are always registered

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Improved Russian translation (#767)
  Remove empty widget width option (#772)
  Fix overlooked reference to App facade instead of local instance
  Use local app instance rather than facade
  Don't register backend permissions in the testing suite
  Type hinting and style tweaks
  Use the model instance returned by getRelationModel directly rather than as a static class reference
  Clone the form model before passing it to the RelationController

# Conflicts:
#	modules/cms/ServiceProvider.php

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Fix AuthManager tests

LukeTowers committed:
Merge branch 'develop' into wip/child-theme

* develop:
  Fix broken return type

LukeTowers committed:
Add ability to specify the cache key of the AutoDatasource

LukeTowers committed:
Set the AutoDatasource's cache key based on the current theme

jaxwilko committed:
Added fix to allow nested AutoDatasource instances to manage their own populateCache

LukeTowers committed:
Merge branch 'develop' into wip/child-theme-autodatasource-caching

* develop:
  Fix issue with autoloading modules (#792)
  Add support for icon picker (#725)
  Classloader improvements (#779)
  Update Markdown content test
  Improved Ukrainian translation (#766)
  Add AJAX form validation to Snowboard (#777)
  Improve theme scaffolding (#778)

LukeTowers committed:
Merge pull request #791 from wintercms/wip/child-theme-autodatasource-caching

[FIX] Added fix to allow nested AutoDatasource instances to manage their own populateCache

jaxwilko committed:
Added support to the theme filter to validate parent theme for assets

LukeTowers commented on modules/cms/classes/Theme.php:
```diff
@@ -402,25 +420,63 @@ public function getFormConfig()
+     */
+    public function assetUrl(?string $path): string
+    {
+        $cacheTime = Config::get('cms.urlCacheTtl', 10);
```
Cache expiration is set in seconds, not minutes, best practice is to always do `now()->addMinutes(10)` instead for most explicit use

jaxwilko replied:
We should probably update this comment then
https://github.com/wintercms/winter/blob/ffa6f99210234f86891def675df69341d30f925c/config/cms.php#L172

And change this value
https://github.com/wintercms/winter/blob/ffa6f99210234f86891def675df69341d30f925c/config/cms.php#L176

jaxwilko replied:
Added this in [1a8fb44](https://github.com/wintercms/winter/pull/726/commits/1a8fb449a9f9989f19ce986b6e442c29cac03ece)

jaxwilko committed:
Added Carbon cache timer

jaxwilko committed:
Removed whitespace

LukeTowers commented on modules/cms/classes/Controller.php:
```diff
@@ -1353,22 +1355,51 @@ public function currentPageUrl($parameters = [], $routePersistence = true)
      * @return string
      */
-    public function themeUrl($url = null)
+    public function themeUrl($url = null): string
+    {
+        return is_array($url)
+            ? $this->themeUrlArray($url)
+            : $this->getTheme()->assetUrl($url);
+    }
+
+    protected function themeUrlArray(array $url): string
```
```suggestion
    public function themeUrl($url = null): string
    {
        return is_array($url)
            ? $this->themeCombineAssets($url)
            : $this->getTheme()->assetUrl($url);
    }

    /**
     * Generates a URL to the AssetCombiner for the provided array of assets
     */
    protected function themeCombineAssets(array $url): string
```

LukeTowers committed:
Update modules/cms/classes/Controller.php

LukeTowers commented on modules/cms/classes/Controller.php:
```diff
@@ -1353,22 +1355,54 @@ public function currentPageUrl($parameters = [], $routePersistence = true)
-            $_url = Url::asset($_url);
+
+            foreach ($url as $file) {
+                if (str_starts_with($file, '@')) {
```
```suggestion
                // Leave Combiner Aliases assets unmodified
                if (str_starts_with($file, '@')) {
```

LukeTowers committed:
Update modules/cms/classes/Controller.php

LukeTowers commented on modules/cms/classes/Controller.php:
```diff
@@ -1353,22 +1355,55 @@ public function currentPageUrl($parameters = [], $routePersistence = true)
+                    }
+                }
+
+                throw new \Exception('Theme URL File not found: ' . $file);
```
```suggestion
                // Skip combining missing assets and log an error
                Log::error("$file could not be found in any of the theme's sources (" . implode(', ', $sources) . ',');
                continue;
```
also test if correct syntax and import the log class

LukeTowers committed:
Update modules/cms/classes/Controller.php

LukeTowers commented on modules/cms/widgets/AssetList.php:
```diff
@@ -405,6 +405,11 @@ protected function getData()
     {
         $assetsPath = $this->getAssetsPath();
 
+        // theme dir does not exist
```
```suggestion
        // theme dir does not exist (i.e. in a child theme without an assets directory
```

LukeTowers committed:
Update modules/cms/widgets/AssetList.php

LukeTowers commented on modules/cms/classes/Theme.php:
```diff
@@ -192,43 +167,80 @@ public static function getActiveThemeCode()
-        if (!File::isDirectory($theme->getPath())) {
-            return self::$activeThemeCache = null;
-        }
+//        if (!File::isDirectory($theme->getPath())) {
+//            return self::$activeThemeCache = null;
+//        }
```
```suggestion
```
Allow virtual themes to exist

LukeTowers committed:
Update modules/cms/classes/Theme.php

jaxwilko committed:
Merge branch 'develop' into wip/child-theme

LukeTowers commented on modules/cms/classes/Theme.php:
```diff
@@ -290,28 +300,26 @@ public static function getEditThemeCode()
-        }
+//        if (!File::isDirectory($theme->getPath())) {
+//            return self::$editThemeCache = null;
+//        }
```
```suggestion
```
Allow virtual themes

LukeTowers commented on modules/cms/classes/Theme.php:
```diff
@@ -332,20 +340,28 @@ public static function all()
+            $sources['database'] = new DbDatasource($this->getDirName(), 'cms_theme_templates');
+        }
+
+        $data = (new AutoDatasource($sources))->selectOne('', 'theme', 'yaml');
```
```suggestion
        // Attempt to load the theme's config file from whatever datasources are available.
        $sources = [
            'filesystem' => new FileDatasource(themes_path($this->getDirName()), App::make('files'))
        ];
        if (static::databaseLayerEnabled()) {
            $sources['database'] = new DbDatasource($this->getDirName(), 'cms_theme_templates');
        }
        $data = (new AutoDatasource($sources))->selectOne('', 'theme', 'yaml');
```

LukeTowers commented on modules/cms/classes/Theme.php:
```diff
@@ -402,25 +417,63 @@ public function getFormConfig()
+     */
+    public function assetUrl(?string $path): string
+    {
+        $cacheTime = now()->addMinutes(Config::get('cms.urlCacheTtl', 10));
+
+        return Cache::remember('winter.cms.assetUrl.' . $path, $cacheTime, function () use ($path) {
```
```suggestion
        $expiresAt = now()->addMinutes(Config::get('cms.urlCacheTtl', 10));
        return Cache::remember('winter.cms.assetUrl.' . $path, $expiresAt, function () use ($path) {
```

LukeTowers replied:
Also please refactor this to use the same approach as the themeUrl() from Cms\Classes\Controller with the foreach over the sources.

LukeTowers commented on modules/cms/classes/Theme.php:
```diff
@@ -471,55 +523,52 @@ public function writeConfig($values = [], $overwrite = false)
 
-        Cache::forget(self::ACTIVE_KEY);
-        Cache::forget(self::EDIT_KEY);
+        if (!$memoryOnly) {
```
```suggestion
        // Sometimes it may be desired to only clear the local cache of the active / edit themes instead of the persistent cache
        if (!$memoryOnly) {
```

LukeTowers commented on modules/cms/classes/Theme.php:
```diff
@@ -528,58 +577,84 @@ public function removeCustomData()
-            $resolver->addDatasource($this->dirName, $datasource);
+            $sources['parent-filesystem'] = new FileDatasource(themes_path($config['parent']), App::make('files'));
         }
+
+        $resolver->addDatasource($this->dirName, new AutoDatasource($sources, 'halcyon-datasource-auto-' . $this->dirName));
```
```suggestion

        $datasource = count($sources) > 1
            ? new AutoDatasource($sources, 'halcyon-datasource-auto-' . $this->dirName)
            : array_shift($sources);
        $resolver->addDatasource($this->dirName, $datasource);
```

LukeTowers committed:
Apply suggestions from code review

jaxwilko committed:
Removed undefined var and replaced with halycon registered event

jaxwilko committed:
Added child theme tests
